File: Encapsulation/CommissionEmployee/CommissionEmployee.java
package Encapsulation.CommissionEmployee;

public class CommissionEmployee {
    private String firstName;
    private String lastName;
    private String socialSecurityNumber;
    private double grossSales; // must be greater than 0.0 on initialization
    private double commissionRate; // must be between 0.0 and 1.0

    public CommissionEmployee (String firstName, String lastName, String socialSecurityNumber, double grossSales, double commissionRate){
        this.firstName = firstName;
        this.lastName = lastName;
        this.socialSecurityNumber = socialSecurityNumber;
        if( grossSales < 0.0){
            throw new IllegalArgumentException("The gross sales should not be less than 0.0.");
            // System.out.println("The grossSales should be greater than 0.0");
            // System.exit(0);
        }
        this.grossSales = grossSales;
        if (commissionRate < 0.0 || commissionRate > 1.0){
            throw new IllegalArgumentException("The commision rate should be between 0.0 and 1.0.");
            // System.out.println("The commissionRate should be between 0.0 and 1.0");
            // System.exit(0);
        }
        this.commissionRate = commissionRate;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getSocialSecurityNumber() {
        return socialSecurityNumber;
    }

    public void setSocialSecurityNumber(String socialSecurityNumber) {
        this.socialSecurityNumber = socialSecurityNumber;
    }

    public double getGrossSales() {
        return grossSales;
    }

    public void setGrossSales(double grossSales) {
        if (grossSales < 0.0){
            throw new IllegalArgumentException("The gross sales should not be less than 0.0.");
            // System.out.println("The grossSales should be greater than 0.0");
            // System.exit(0);
        }
        this.grossSales = grossSales;
    }

    public double getCommissionRate() {
        return commissionRate;
    }

    public void setCommissionRate(double commissionRate) {
        if (commissionRate < 0.0 || commissionRate > 1.0){
            throw new IllegalArgumentException("The commision rate should be between 0.0 and 1.0.");
            // System.out.println("The commissionRate should be between 0.0 and 1.0");
            // System.exit(0);
        }
        this.commissionRate = commissionRate;
    }

    public double earnings(){
        return this.grossSales * this.commissionRate;
    }

    // to display the employee's detail we have to first override its implementation since the toString method on default returns  the Class Name, "@" and Hash Code( representation of the class)
    @Override
    public String toString(){
        return "The Details of Employee; \n Name: " + firstName + " " + lastName + "\n Social security number: "+ socialSecurityNumber + "\n GrossSales: " + grossSales + "\n CommissionRate: " + commissionRate;
    }
    public static void main(String[] args) {
        CommissionEmployee Solomon = new CommissionEmployee("Solomon", "Zigla", "22012447", 50.0, 0.5);
        System.out.println(Solomon.toString());

        // updating employee's grossSale and commissionRate
        Solomon.setCommissionRate(0.7);
        Solomon.setGrossSales(150);

        // Displaying updated details
        System.out.println(Solomon.toString());
        System.out.println("Updated commission rate: " + Solomon.getCommissionRate());
        System.out.println("Updated gross sales: " + Solomon.getGrossSales());

        // calculating employee's earing 
        double amount = Solomon.earnings();
        System.out.println(Solomon.getFirstName() + " " + Solomon.getLastName() + " makes a total earning of " + amount);

        // testing validation with invalid values
        Solomon.setCommissionRate(-0.2);
        Solomon.setGrossSales(-5);
        
    }
}

File: python_lab_tasks/CommissionEmployee.py
class CommissionEmployee:
    def __init__(self, first_name, last_name, social_security_number, gross_sales, commission_rate):
        self.__first_name = first_name
        self.__last_name = last_name
        self.__social_security_number = social_security_number
        if gross_sales >= 0.0:
            self.__gross_sales = gross_sales
        else:
            raise ValueError("Gross sales must be >= 0.0")

        if 0.0 <= commission_rate <= 1.0:
            self.__commission_rate = commission_rate
        else:
            raise ValueError("Commission rate must be between 0.0 and 1.0")

    def get_first_name(self):
        return self.__first_name

    def set_first_name(self, value):
        self.__first_name = value

    def get_last_name(self):
        return self.__last_name

    def set_last_name(self, value):
        self.__last_name = value

    def get_social_security_number(self):
        return self.__social_security_number

    def set_social_security_number(self, value):
        self.__social_security_number = value

    def get_gross_sales(self):
        return self.__gross_sales

    def set_gross_sales(self, value):
        if value >= 0.0:
            self.__gross_sales = value
        else:
            raise ValueError("Gross sales must be >= 0.0")

    def get_commission_rate(self):
        return self.__commission_rate

    def set_commission_rate(self, value):
        if 0.0 <= value <= 1.0:
            self.__commission_rate = value
        else:
            raise ValueError("Commission rate must be between 0.0 and 1.0")

    def earnings(self):
        return self.__gross_sales * self.__commission_rate

    def display_employee_details(self):
        return (f"CommissionEmployee: {self.__first_name} {self.__last_name}\n"
                f"Social Security Number: {self.__social_security_number}\n"
                f"Gross Sales: {self.__gross_sales}\n"
                f"Commission Rate: {self.__commission_rate}")

# Implementation Tasks
if __name__ == "__main__":
    employee = CommissionEmployee("Solomon", "Zigla", "22012447", 700.0, 0.1)
    employee.set_gross_sales(900.0)
    employee.set_commission_rate(0.15)

    print(employee.display_employee_details())
    print(f"Earnings: {employee.earnings()}")
